src/util/simplify_expr_pointer.cpp

```diff
@@ -421,48 +422,69 @@ simplify_exprt::simplify_pointer_offset(const unary_exprt &expr)
 simplify_exprt::resultt<> simplify_exprt::simplify_inequality_address_of(
   const binary_relation_exprt &expr)
 {
+  // the operands of the relation are both either one of
+  // a) an address_of_exprt
+  // b) a typecast_exprt with an address_of_exprt operand
+
   PRECONDITION(expr.id() == ID_equal || expr.id() == ID_notequal);
 
   exprt tmp0=expr.op0();
+
+  // skip over the typecast
   if(tmp0.id()==ID_typecast)
-    tmp0=expr.op0().op0();
-  if(tmp0.op0().id()==ID_index &&
-     to_index_expr(tmp0.op0()).index().is_zero())
-    tmp0=address_of_exprt(to_index_expr(tmp0.op0()).array());
+    tmp0 = to_typecast_expr(tmp0).op();
+
+  PRECONDITION(tmp0.id() == ID_address_of);
```

Comment:
That's also checked in `to_address_of_expr`

Reply:
yes, left deliberately, to document the contract of this function.